File: number/factor/factor.py
#!/usr/bin/env python
#-*- encoding: UTF-8-*-
"""
@desc
质因数分解python语言示例
一个偶数可以 分解为若干个质数相乘。
"""
import time

def factorize1(num):
  i = 2
  result = []
  while (num > 1):
    while (num % i == 0):
      result.append(i)
      num = num / i
    i = i + 1
  print(result)

def factorize(num):
  i = 2
  result = []
  while (num >= i * i):
    while (num % i == 0):
      result.append(i)
      num = num / i
    i = i + 1
  if (num > 1):
    result.append(num)
  return result

if __name__ == '__main__':

    # 1. factorize1
    num = 20
    print(num)
    start_time = time.time()
    arr = factorize1(num)
    print("time:" + str((time.time() - start_time) * 1000) + " ms")

    # 2. factorize2
    num = 20
    print("%d =") %(num),
    start_time = time.time()
    arr = factorize(num)
    size = len(arr)
    for i in range(size):
      print(arr[i]),
      if (i < size - 1):
        print("*"),

    print("\r\ntime:" + str((time.time() - start_time) * 1000) + " ms")

"""
jarry@jarrys-MacBook-Pro factor % python --version 
Python 2.7.16
jarry@jarrys-MacBook-Pro factor % python factor.py
20
[2, 2, 5]
time:0.014066696167 ms
20 = 2 * 2 * 5 
time:0.00905990600586 ms
*/
"""

File: number/factor/Factor.java

/**
 * @desc
 * 质因数分解Java语言示例
 */

import java.util.*;

class Factor {

  public static List<Integer> factorize(int num) {
    int i = 2;
    List<Integer> result = new ArrayList<>();
    while (num > i * i) {
      while (num % i == 0) {
        result.add(i);
        num /= i;
      }
      i++;
    }
    if (num > 1) {
      result.add(num);
    }
    return result;
  }

  public static void factorize2(int num) {
    int i = 2;
    while (num > 1) {
      while (num % i == 0) {
        System.out.print(i);
        if (num != i) {
          System.out.print(" * ");
        }
        num /= i;
      }
      i++;
    }
  }

  public static void main(String args[]) {
    int num = 20;
    
    long startTime;
    // test1
    System.out.println("start:" + num);
    startTime = System.currentTimeMillis();
    List<Integer> result = Factor.factorize(num);
    for (int i = 0; i < result.size(); i++) {
      System.out.print(result.get(i));
      if (i < result.size() - 1) {
        System.out.print(" * ");
      }
    }
    System.out.println("\r\ntime:" + (System.currentTimeMillis() - startTime) + " ms.");

    // test 2
    startTime = System.currentTimeMillis();
    Factor.factorize2(num);
    System.out.println("\r\ntime:" + (System.currentTimeMillis() - startTime) + " ms.");
  }

}

/**
jarry@jarrys-MacBook-Pro factor % java --version
java 14.0.1
Java(TM) SE Runtime Environment (build 14.0.1+7)
Java HotSpot(TM) 64-Bit Server VM (build 14.0.1+7, mixed mode, sharing)
jarry@jarrys-MacBook-Pro factor % javac Factor.java
jarry@jarrys-MacBook-Pro factor % java Factor
start:20
2 * 2 * 5
time:0 ms.
2 * 2 * 5
time:1 ms.
*/